File: code/java/rest/client/src/main/java/com/taskapi/client/TaskApiClient.java
package com.taskapi.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import lombok.extern.slf4j.Slf4j;
import okhttp3.*;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.*;
import java.util.concurrent.TimeUnit;

@Slf4j
public class TaskApiClient {
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final String baseUrl;
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    
    public TaskApiClient(String baseUrl) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .writeTimeout(10, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();
        
        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new JavaTimeModule());
        this.objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
    }
    
    public List<Map<String, Object>> listTasks(Integer pageSize, String pageToken, 
                                                String status, String assignedTo, 
                                                List<String> tags, String sortOrder) throws IOException {
        HttpUrl.Builder urlBuilder = HttpUrl.parse(baseUrl + "/api/v1/tasks").newBuilder();
        
        if (pageSize != null) {
            urlBuilder.addQueryParameter("pageSize", String.valueOf(pageSize));
        }
        if (pageToken != null) {
            urlBuilder.addQueryParameter("pageToken", pageToken);
        }
        if (status != null) {
            urlBuilder.addQueryParameter("status", status);
        }
        if (assignedTo != null) {
            urlBuilder.addQueryParameter("assignedTo", assignedTo);
        }
        if (tags != null && !tags.isEmpty()) {
            urlBuilder.addQueryParameter("tags", String.join(",", tags));
        }
        if (sortOrder != null) {
            urlBuilder.addQueryParameter("sortOrder", sortOrder);
        }
        
        Request request = new Request.Builder()
                .url(urlBuilder.build())
                .get()
                .build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Unexpected response code: " + response);
            }
            
            Map<String, Object> result = objectMapper.readValue(response.body().string(), Map.class);
            return (List<Map<String, Object>>) result.get("items");
        }
    }
    
    public Map<String, Object> getTask(String taskId) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + "/api/v1/tasks/" + taskId)
                .get()
                .build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            if (response.code() == 404) {
                return null;
            }
            if (!response.isSuccessful()) {
                throw new IOException("Unexpected response code: " + response);
            }
            
            return objectMapper.readValue(response.body().string(), Map.class);
        }
    }
    
    public Map<String, Object> createTask(String title, String description, String priority,
                                           List<String> tags, String assignedTo) throws IOException {
        Map<String, Object> task = new HashMap<>();
        task.put("title", title);
        if (description != null) task.put("description", description);
        if (priority != null) task.put("priority", priority.toUpperCase());
        if (tags != null) task.put("tags", tags);
        if (assignedTo != null) task.put("assignedTo", assignedTo);
        
        String json = objectMapper.writeValueAsString(task);
        RequestBody body = RequestBody.create(json, JSON);
        
        Request request = new Request.Builder()
                .url(baseUrl + "/api/v1/tasks")
                .post(body)
                .build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to create task: " + response);
            }
            
            return objectMapper.readValue(response.body().string(), Map.class);
        }
    }
    
    public Map<String, Object> updateTask(String taskId, Map<String, Object> updates) throws IOException {
        String json = objectMapper.writeValueAsString(updates);
        RequestBody body = RequestBody.create(json, JSON);
        
        Request request = new Request.Builder()
                .url(baseUrl + "/api/v1/tasks/" + taskId)
                .put(body)
                .build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            if (response.code() == 404) {
                return null;
            }
            if (!response.isSuccessful()) {
                throw new IOException("Failed to update task: " + response);
            }
            
            return objectMapper.readValue(response.body().string(), Map.class);
        }
    }
    
    public Map<String, Object> updateTaskStatus(String taskId, String status) throws IOException {
        Map<String, String> statusUpdate = new HashMap<>();
        statusUpdate.put("status", status.toUpperCase());
        
        String json = objectMapper.writeValueAsString(statusUpdate);
        RequestBody body = RequestBody.create(json, JSON);
        
        Request request = new Request.Builder()
                .url(baseUrl + "/api/v1/tasks/" + taskId + "/status")
                .patch(body)
                .build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            if (response.code() == 404) {
                return null;
            }
            if (!response.isSuccessful()) {
                throw new IOException("Failed to update task status: " + response);
            }
            
            return objectMapper.readValue(response.body().string(), Map.class);
        }
    }
    
    public boolean deleteTask(String taskId) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + "/api/v1/tasks/" + taskId)
                .delete()
                .build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            return response.code() == 204;
        }
    }
    
    public void close() {
        httpClient.dispatcher().executorService().shutdown();
        httpClient.connectionPool().evictAll();
    }
    
    public static void main(String[] args) {
        TaskApiClient client = new TaskApiClient("http://localhost:8080");
        
        try {
            // Create a task
            log.info("Creating a new task...");
            Map<String, Object> newTask = client.createTask(
                    "Test Java REST Client",
                    "Testing the Java REST client implementation",
                    "HIGH",
                    Arrays.asList("test", "java", "rest"),
                    "dev-team"
            );
            log.info("Created task: {}", newTask);
            
            String taskId = (String) newTask.get("id");
            
            // Get the task
            log.info("\nRetrieving task {}...", taskId);
            Map<String, Object> task = client.getTask(taskId);
            log.info("Retrieved task: {}", task);
            
            // Update task status
            log.info("\nUpdating task status to IN_PROGRESS...");
            Map<String, Object> updated = client.updateTaskStatus(taskId, "IN_PROGRESS");
            log.info("Updated task: {}", updated);
            
            // List all tasks
            log.info("\nListing all tasks...");
            List<Map<String, Object>> tasks = client.listTasks(10, null, null, null, null, "created_desc");
            for (Map<String, Object> t : tasks) {
                log.info("[{}] {} - {}", t.get("status"), t.get("title"), t.get("id"));
            }
            
            // Delete the task
            log.info("\nDeleting task {}...", taskId);
            boolean deleted = client.deleteTask(taskId);
            log.info("Task deleted: {}", deleted);
            
        } catch (IOException e) {
            log.error("Error communicating with API", e);
        } finally {
            client.close();
        }
    }
}

File: code/kotlin/rest/client/src/main/kotlin/com/taskapi/client/TaskApiClient.kt
package com.taskapi.client

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.runBlocking
import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
enum class TaskStatus {
    PENDING, IN_PROGRESS, COMPLETED, CANCELLED, ARCHIVED
}

@Serializable
enum class TaskPriority {
    LOW, MEDIUM, HIGH, CRITICAL
}

@Serializable
data class Task(
    val id: String,
    val title: String,
    val description: String? = null,
    val status: TaskStatus,
    val priority: TaskPriority,
    val tags: List<String> = emptyList(),
    val assignedTo: String? = null,
    val dueDate: Instant? = null,
    val createdAt: Instant,
    val updatedAt: Instant,
    val createdBy: String? = null,
    val updatedBy: String? = null
)

@Serializable
data class CreateTaskRequest(
    val title: String,
    val description: String? = null,
    val priority: TaskPriority = TaskPriority.MEDIUM,
    val tags: List<String> = emptyList(),
    val assignedTo: String? = null,
    val dueDate: Instant? = null
)

@Serializable
data class UpdateTaskRequest(
    val title: String? = null,
    val description: String? = null,
    val status: TaskStatus? = null,
    val priority: TaskPriority? = null,
    val tags: List<String>? = null,
    val assignedTo: String? = null,
    val dueDate: Instant? = null
)

@Serializable
data class UpdateStatusRequest(
    val status: TaskStatus
)

@Serializable
data class TasksResponse(
    val items: List<Task>,
    val pageSize: Int,
    val nextPageToken: String? = null,
    val totalCount: Int
)

class TaskApiClient(private val baseUrl: String) {
    private val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json {
                prettyPrint = true
                isLenient = true
                ignoreUnknownKeys = true
            })
        }
        install(Logging) {
            level = LogLevel.INFO
        }
        install(HttpTimeout) {
            requestTimeoutMillis = 30000
            connectTimeoutMillis = 10000
            socketTimeoutMillis = 30000
        }
        defaultRequest {
            header(HttpHeaders.ContentType, ContentType.Application.Json)
        }
    }

    suspend fun listTasks(
        pageSize: Int = 20,
        pageToken: String? = null,
        status: TaskStatus? = null,
        assignedTo: String? = null,
        tags: List<String>? = null,
        sortOrder: String? = null
    ): TasksResponse {
        val response: HttpResponse = client.get("$baseUrl/api/v1/tasks") {
            parameter("pageSize", pageSize)
            pageToken?.let { parameter("pageToken", it) }
            status?.let { parameter("status", it.name) }
            assignedTo?.let { parameter("assignedTo", it) }
            tags?.let { parameter("tags", it.joinToString(",")) }
            sortOrder?.let { parameter("sortOrder", it) }
        }
        return response.body()
    }

    suspend fun getTask(taskId: String): Task? {
        return try {
            val response: HttpResponse = client.get("$baseUrl/api/v1/tasks/$taskId")
            if (response.status == HttpStatusCode.NotFound) {
                null
            } else {
                response.body()
            }
        } catch (e: ClientRequestException) {
            if (e.response.status == HttpStatusCode.NotFound) {
                null
            } else {
                throw e
            }
        }
    }

    suspend fun createTask(
        title: String,
        description: String? = null,
        priority: TaskPriority = TaskPriority.MEDIUM,
        tags: List<String> = emptyList(),
        assignedTo: String? = null
    ): Task {
        val request = CreateTaskRequest(
            title = title,
            description = description,
            priority = priority,
            tags = tags,
            assignedTo = assignedTo
        )

        val response: HttpResponse = client.post("$baseUrl/api/v1/tasks") {
            setBody(request)
        }
        return response.body()
    }

    suspend fun updateTask(
        taskId: String,
        updates: UpdateTaskRequest
    ): Task? {
        return try {
            val response: HttpResponse = client.put("$baseUrl/api/v1/tasks/$taskId") {
                setBody(updates)
            }
            response.body()
        } catch (e: ClientRequestException) {
            if (e.response.status == HttpStatusCode.NotFound) {
                null
            } else {
                throw e
            }
        }
    }

    suspend fun updateTaskStatus(taskId: String, status: TaskStatus): Task? {
        return try {
            val request = UpdateStatusRequest(status)
            val response: HttpResponse = client.patch("$baseUrl/api/v1/tasks/$taskId/status") {
                setBody(request)
            }
            response.body()
        } catch (e: ClientRequestException) {
            if (e.response.status == HttpStatusCode.NotFound) {
                null
            } else {
                throw e
            }
        }
    }

    suspend fun deleteTask(taskId: String): Boolean {
        return try {
            val response: HttpResponse = client.delete("$baseUrl/api/v1/tasks/$taskId")
            response.status == HttpStatusCode.NoContent
        } catch (e: ClientRequestException) {
            if (e.response.status == HttpStatusCode.NotFound) {
                false
            } else {
                throw e
            }
        }
    }

    fun close() {
        client.close()
    }
}

suspend fun main() {
    val client = TaskApiClient("http://localhost:8080")

    try {
        // Create a task
        println("Creating a new task...")
        val task = client.createTask(
            title = "Test Kotlin REST Client",
            description = "Testing the Kotlin REST client implementation",
            priority = TaskPriority.HIGH,
            tags = listOf("test", "kotlin", "rest"),
            assignedTo = "dev-team"
        )
        println("Created task: ${task.id} - ${task.title}")

        // Get the task
        println("\nRetrieving task ${task.id}...")
        val retrieved = client.getTask(task.id)
        if (retrieved != null) {
            println("Retrieved task: ${retrieved.title} - Status: ${retrieved.status}")
        }

        // Update task status
        println("\nUpdating task status to IN_PROGRESS...")
        val updated = client.updateTaskStatus(task.id, TaskStatus.IN_PROGRESS)
        if (updated != null) {
            println("Updated task status: ${updated.status}")
        }

        // List all tasks
        println("\nListing all tasks...")
        val tasks = client.listTasks(pageSize = 10, sortOrder = "created_desc")
        tasks.items.forEach { t ->
            println("[${t.status}] ${t.title} - ${t.id}")
        }
        println("Total tasks: ${tasks.totalCount}")

        // Update task details
        println("\nUpdating task details...")
        val updateRequest = UpdateTaskRequest(
            description = "Updated: Testing the Kotlin REST client implementation with Ktor",
            priority = TaskPriority.MEDIUM,
            tags = listOf("test", "kotlin", "rest", "ktor")
        )
        val updatedTask = client.updateTask(task.id, updateRequest)
        if (updatedTask != null) {
            println("Updated task: ${updatedTask.description}")
        }

        // Delete the task
        println("\nDeleting task ${task.id}...")
        val deleted = client.deleteTask(task.id)
        println("Task deleted: $deleted")

    } catch (e: Exception) {
        println("Error: ${e.message}")
        e.printStackTrace()
    } finally {
        client.close()
    }
}